core/src/main/java/io/grpc/ServerInterceptors.java

```diff
@@ -288,4 +290,81 @@ public void onMessage(WReqT message) {
       }
     };
   }
+
+  /**
+   * A class that intercepts uncaught exceptions of type {@link StatusRuntimeException} and handles
+   * them by closing the {@link ServerCall}, and transmitting the exception's details to the client.
+   *
+   * <p>Without this interceptor, gRPC will strip all details and close the {@link ServerCall} with
+   * a generic {@link Status#UNKNOWN} code.
+   *
+   * <p>Security warning: the {@link Status} and {@link Metadata} may contain sensitive server-side
+   * state information, and generally should not be sent to clients. Only install this interceptor
+   * if all clients are trusted.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2189")
+  public static final ServerInterceptor STATUSRUNTIMEXCEPTION_TRANSMITTING_INTERCEPTOR =
+      new ServerInterceptor() {
+        @Override
+        public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+            ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+          final ServerCall<ReqT, RespT> syncServerCall =
+              new SynchronizedServerCall<ReqT, RespT>(call);
+          ServerCall.Listener<ReqT> listener = next.startCall(syncServerCall, headers);
+          return new SimpleForwardingServerCallListener<ReqT>(listener) {
+            @Override
+            public void onMessage(ReqT message) {
+              try {
+                super.onMessage(message);
+              } catch (StatusRuntimeException t) {
+                closeWithException(t);
+              }
+            }
+
+            @Override
+            public void onHalfClose() {
+              try {
+                super.onHalfClose();
+              } catch (StatusRuntimeException t) {
+                closeWithException(t);
+              }
+            }
+
+            @Override
+            public void onCancel() {
+              try {
+                super.onCancel();
+              } catch (StatusRuntimeException t) {
+                closeWithException(t);
+              }
+            }
+
+            @Override
+            public void onComplete() {
+              try {
+                super.onComplete();
+              } catch (StatusRuntimeException t) {
+                closeWithException(t);
+              }
+            }
+
+            @Override
+            public void onReady() {
+              try {
+                super.onReady();
+              } catch (StatusRuntimeException t) {
+                closeWithException(t);
```

Comment:
Per offline discussion, the deadlock issue with in-process transport is legit, while not the fault of this PR. I have filed #3084

However, even without #3084, using `synchronized` to make thread-safe calls to `ServerCall` and `ClientCall` is still prone to deadlock when in-process transport is used with direct executor:

- Thread1
  1. `SynchronizedClientCall.sendMessage()` (locks `SynchronizedClientCall.this`)
  1. `ServerCall.Listener` calls into application code, which calls `SynchronizedServerCall.close()` (locks `SynchronizedServerCall.this`)

- Thread2
  1. `SynchronizedServerCall.sendMessage()` (locks `SynchronizedServerCall.this`)
  1. `ClientCall.Listener` calls into application code, which calls `SynchronizedClientCall.close()` (locks `SynchronizedClientCall.this`)

This time, it's the two synchronized calls that are involved in the deadlock.

This can be considered application-level code, and if application does that and got deadlock, it's the application's fault. Application shouldn't do it, and it means **we should not do it either**.

Like the proposed solution for #3084, we could also use `ChannelExecutor` instead to provide the thread-safety. `ChannelExecutor` should probably be renamed.